### ISSUE TITLE
Improved Notifications (v1) Spike

### ISSUE THREAD

**Issue by sachben91** [hicommonwealth/commonwealth#6545]
## Context 
Once progressive web app is enabled we want to be able to send push notifications to users. These would include notifications about discussions and also user engagement notifications in the future. 

[Please read the PRD and leave comments before proceeding.](https://www.notion.so/buildcommon/Push-notifications-V1-145890b638e34ce589a718cf119ccdae)

## Project Owners 
PM: @sachben91 
Design: Jared Bell 

## Product questions for spike 
- what is the effort involved in setting up the infrastructure for sending push notifications once pwa is enabled? 
- would setting up notifications on IoS and Android be separate efforts? 
- Are there notifications in the PRD that are not feasible?

**Comment by jnaviask** [comment]
Instead of PRD, 1 eng + @timolegros on 30min call with Sachin. After call, 2 hour timebox to complete feasibility studies + writeup.

**Comment by timolegros** [comment]
Noting here that I will complete the feasibility analysis + write up with @Rotorsoft after the event storming on Feb 8.

**Comment by timolegros** [comment]
As discussed with Sachin and the team, since the PRD includes changes for all existing notification channels (in-app, email, push, webhooks) this is a feasibility analysis that covers all of these changes rather than just push notifications.

# Summary of Requirements
## Notification Types
- User opens PWA for the first time
- Update on stakes burned in the community every 8 hours
- Update on stakes minted in the community every 8 hours
- New thread
- New comment (on root thread or as a reply to another comment)
- New mentions
## Channels
- In-app
- Push (IOS/Android)
- Batch email

Note: Webhooks are not included in the channels here since this PRD focuses on user notifications while webhooks are community notifications. That said, if we choose to use an external system to implement improved notifications, we should ensure that webhooks integrate nicely with it.

# Next Steps
1. Discuss with platform team the approach we want to go with
2. Create an overarching architecture bucket ticket
3. Create appropriate child (potentially bucket) tickets for each set of changes (models, channels, etc)

**Comment by timolegros** [comment]
As discussed with Sachin and the team, since the PRD includes changes for all existing notification channels (in-app, email, push, webhooks) this is a feasibility analysis that covers all of these changes rather than just push notifications.

# Notifications by the numbers
### Events
Over the last 3 months we have generated 15,496 events.
  - `new-thread-creation`: 1068
  - `new-comment-creation`: 4203
  - `new-mention`: 0 (these are currently broken so expect more than 0 in the new system)
  - `new-reaction`: 3074
  - `chain-event`: 164
  - `new-collaboration`: 0
  - `thread-edit`: 1082
  - `comment-edit`: 622
  - `snapshot-proposal`: 5283

### User Notifications
Over the last 3 months we have created 178,522 user notifications based on the events
  - `new-thread-creation`: 124126
  - `new-comment-creation`: 25393
  - `new-mention`: 0 (these are currently broken so expect more than 0 in the new system)
  - `new-reaction`: 5721
  - `chain-event`: 1099
  - `new-collaboration`: 0
  - `thread-edit`: 0 (only used for feeds)
  - `comment-edit`: 0 (only used for feeds)
  - `snapshot-proposal`: 22180

# New Notification Requirements
![Screenshot from 2024-02-12 21-34-31](https://github.com/hicommonwealth/commonwealth/assets/62490329/92fa1289-eee6-41d7-989d-a4420e24a82c)
- Must be able to construct/integrate new notification templates, workflows, and destinations (e.g. Slack, Whatsapp) without significantly altering the codebase.
- Must directly (or via integrations) support all notification channels (in-app, mobile push - IOS and Android, email, and webhooks)
- Must have a NodeJS SDK for interacting with the API and a React package for building user-facing components.

Note: Webhooks are not included in the channels here since this PRD focuses on user notifications while webhooks are community notifications. That said, if we choose to use an external system to implement improved notifications, we should ensure that webhooks integrate nicely with it.

# Approaches/Comparison
## Courier (vs Knock)
### Pros
- Established in 2019 (closed 35$ million series B in 2022). Courier is a safer long-term pick if considering stability/longevity.
  - Knock was established in 2022. Knock is still very new and has much less runway than Courier (closed 2.7$ million funding round in 2022).
- Larger selection of integrations (48)
  - Knock supports approximately 23 integrations
- Larger selection and better-maintained SDKs. The Courier Node SDK has better support: https://github.com/trycourier/courier-node compared to the Knock Node SDK https://github.com/knocklabs/knock-node
- User-defined notification digests.
  - Knock only allows developers to define digests via workflows.
- Support for incrementally building a bulk job that is eventually triggered (e.g. build list of users over time and send at the end of the week)
  - On Knock we would need to build/maintain the list of users ourselves
- More robust multi-tenancy (relevant for custom community or user branding)

Note: everything Knock can do, Courier can do as well.

### Cons
- No options for self-hosting 
  - Knock provides self-hosting but this is mostly for enterprises that must abide by strict regulations or have very complex needs.
- I found that the Courier docs are not as thorough or organized as Knock. This is mitigated by the fact that the Courier community (and company) is larger so responsiveness on GitHub issues or Slack messages with the team is likely better than Knock.

## Novu
Reviewed Novu and it quickly became clear that Novu is a great open-source project but it is not nearly as feature-full as Knock and Courier despite having similar pricing for hosted versions.

## In-house
There is no need to reinvent the wheel. When using an external service, a PM can modify a workflow to add a whole new destination e.g. Whatsapp without ever having to change any code. Without an external service, we would need to implement our own adapter to send notifications to the new destination. The additional cost of development for such integrations greatly outweighs the cost of external services.

To implement an in-house notifications system we would need to set up a pipeline on AWS as detailed here: https://www.notion.so/buildcommon/Notifications-Pipeline-Design-92bb4f9e2a6344ba9b4b8f6653eea9d0

# Cost
Courier, Knock, and Novu all have the same/similar pricing. Pricing is not a differentiating factor between external services. Regardless of the external service we use, we are looking at $0.005 per notification which comes out to roughly $890 (178k * 0.005) per month. If we ignore the initial development cost of integrating an external service, $890 is roughly equivalent to 2 days of work for an IC. Creating any of the features that these systems provide for us in-house would likely take much more than 2 days so if we expect any type of growth in terms of notifications workflows or notification destinations, an external system is cost-efficient. Additionally, an in-house system would require much more maintenance thus increasing the cost even further.

# Next Steps
1. Discuss with the platform team the approach we want to go with
2. Decide whether we want to keep track of user notification/subscription data ourselves as well as on the external service.
3. Create an overarching architecture bucket ticket
4. Create appropriate child (potentially bucket) tickets for each set of changes (models, channels, etc)

**Comment by timolegros** [comment]
Takeaways from a conversation with Roger:

- We need to persist a source of truth about events that have occurred as well as user preferences.
  - This reduces our coupling with any external service and ensures we can easily switch out the notification provider. It also ensures that we can use the events for other purposes e.g. analytics.
  - Concretely, this means we will keep the `Notifications`, and `Subscriptions` (this one will change significantly) tables but remove the `NotificationsReads` table.
- We should implement the Transactional Outbox pattern for emitting events. Generally speaking, this means when we create an object like a thread in the `Threads` table, we also need to insert the JSON data of that thread record in an `Outbox` table. Events from the Outbox table are then read by a message relayer which pushes the message to a queue. At the other end of the queue, we would need a worker that processes the events and in our case makes a request to the Courier or Knock API.
 - Because the full command framework isn't implemented yet, we should utilize Sequelize hooks to implement the 'publish' step of the command framework only for the events that are relevant for notifications.

**Comment by jnaviask** [comment]
Really fantastic analysis. All points make sense, and I really appreciate seeing the numbers. Some follow up questions, for prioritization/scoping/decision-making:

- what would courier cost be as we scale to 1mm notifications/mo from 100k? How about 10mm? What would the cost be with other providers, if less? ~$1k/mo is definitely acceptable but as we scale we should ensure that costs don't grow linearly, or we'll be paying more for notifications than for e.g. web hosting.
- what's the tentative list of steps required to implement the new system? How many points approx per step?
- what existing components would be deprecated / could be removed following the implementation? Are there any features that we currently support which would be difficult to reimplement using the new system approach?

**Comment by timolegros** [comment]
> Really fantastic analysis. All points make sense, and I really appreciate seeing the numbers. Some follow up questions, for prioritization/scoping/decision-making:
> 
> * what would courier cost be as we scale to 1mm notifications/mo from 100k? How about 10mm? What would the cost be with other providers, if less? ~$1k/mo is definitely acceptable but as we scale we should ensure that costs don't grow linearly, or we'll be paying more for notifications than for e.g. web hosting.
> * what's the tentative list of steps required to implement the new system? How many points approx per step?
> * what existing components would be deprecated / could be removed following the implementation? Are there any features that we currently support which would be difficult to reimplement using the new system approach?

1. It's hard to say what the cost would be per notification above several hundred thousand. The $0.005 per notification I am referencing is the 2nd tier plan (they all have 3 -> free tier, contract, and enterprise) so if we go with an enterprise plan (which we probably should after a certain amount of time) we can probably get some discount per notification e.g. $0.004 per notification but minimum 1mm notifications. Roger and I also discussed some options to reduce the number of notifications emitted i.e. batching, making notifications more relevant to users (meaning not every thread is a notification but only X threads with a certain rank), etc. In other words, we probably already emit more notifications than we need to especially when we compare to how Reddit/Twitter emits notifications.
2. I will have a better answer for this tomorrow since I will create the bucket tickets with Roger but since this will be event-driven architecture with the transactional outbox pattern, there is additional work to be done on top of integrating the external provider i.e. set up 2 workers, queue, hooks, etc. Definitely looking at a 13-pointer overall.
3. `EmitNotifications` script, `NotificationsReads` table, `Notifications` table, the Clean DB script, all notification routes, email scripts (including the Heroku scheduler jobs), and webhook scripts would all be deprecated (not an exhaustive list but covers the major areas). Thus far I haven't come across anything that would be difficult to reimplement using the new system but I also haven't built an MVP with the workflow tools the providers have so I can't say for certain that there is absolutely nothing. One thing to note is that the feeds will be based on the outbox events table rather than the notifications table so we will need to think carefully about how we do that - but it isn't directly related to an external provider. That said, the goal should be to eventually show algorithmically ranked data on the feeds rather than just replicating whatever events/notifications have been emitted.

**Comment by dillchen** [comment]
TLDR: considering future costs at 10x is not a blocker.

-----
This all looks good, we will have a good problem on our hands at 10x scale.

I am not worried about the cost at 10x, $10k per month, without any discount 

I'll assume we get enterprise pricing so 10x user scale comes down to smth like 5-7x in costs.

That's still 2w eng time per month.

At that point in time, we will either 1) go in house. 2) stay with them

I believe Roger's point allows us to not have platform lock-in. So this is derided from my perspective.

**Comment by rbennettcw** [comment]
> Takeaways from a conversation with Roger:
> 
> * We need to persist a source of truth about events that have occurred as well as user preferences.
>   
>   * This reduces our coupling with any external service and ensures we can easily switch out the notification provider. It also ensures that we can use the events for other purposes e.g. analytics.
>   * Concretely, this means we will keep the `Subscriptions` table (will change significantly) but remove the `NotificationsReads` and `Notifications` table (replaced by the Outbox).
> * We should implement the [Transactional Outbox](https://microservices.io/patterns/data/transactional-outbox.html) pattern for emitting events. Generally speaking, this means when we create an object like a thread in the `Threads` table, we also need to insert the JSON data of that thread record in an `Outbox` table. Events from the Outbox table are then read by a message relayer which pushes the message to a queue. At the other end of the queue, we would need a worker that processes the events and in our case makes a request to the Courier or Knock API.
> * Because the full command framework isn't implemented yet, we should utilize Sequelize hooks to implement the 'publish' step of the command framework only for the events that are relevant for notifications.

Everything looks good.

One note about Sequelize hooks– they only trigger via the ORM, but not for raw SQL queries. Should be fine considering most CUD operations use the ORM while the raw SQL is mostly used for reads. But it's something to double-check for each model will have Sequelize hooks added.

**Comment by Rotorsoft** [comment]
We will discuss implementation details soon. There are many patterns to populate the outbox, the key requirement here is that the state mutation and outbox operations must be within the same ACID transaction. 

Possible solutions:
- Sequelize hooks
- `pg` triggers - if we conceptualize events as generic state changes (simple)
- Encapsulate the transaction requirement in a framework layer that emits domain events - more DDD but requires a little bit of work 

Another point to consider is the structural relationships we need to synchronize with the notifications provider in order to allow their scheduler to produce scoped notification events - thinking about "notify community level events to all subscribed users that are member of the community", or when users subscribe to specific entities... hoping we never get to think about gated notifications 😆
